src/main/java/net/minestom/server/entity/EntityQuery.java

```diff
@@ -0,0 +1,169 @@
+
+    sealed interface Condition<T> extends BiPredicate<Point, T> permits EntityQueryImpl.ConditionImpl {
+        @NotNull Property<T> property();
+
+        @Override
+        boolean test(@NotNull Point origin, @NotNull T value);
+
+        static <T> Condition<T> equalsCondition(Property<T> property, T value) {
+            return __condition(property, (point, t) -> t.equals(value));
+        }
+
+        static <T> Condition<T> notEqualsCondition(Property<T> property, T value) {
+            return __condition(property, (origin, t) -> !value.equals(t));
+        }
+
+        static <T extends Number> Condition<T> rangeCondition(Property<T> property, Range<T> range) {
+            return __condition(property, (origin, t) -> switch (range) {
+                case Range.Byte aByte -> aByte.inRange((Byte) t);
+                case Range.Double aDouble -> aDouble.inRange((Double) t);
+                case Range.Float aFloat -> aFloat.inRange((Float) t);
+                case Range.Int anInt -> anInt.inRange((Integer) t);
+                case Range.Long aLong -> aLong.inRange((Long) t);
+                case Range.Short aShort -> aShort.inRange((Short) t);
+            });
+        }
+
+        static Condition<Point> chunkRangeCondition(int radius) {
+            return __condition(COORD, (origin, coord) -> {
+                final int originChunkX = origin.chunkX();
+                final int originChunkZ = origin.chunkZ();
+                final int coordChunkX = coord.chunkX();
+                final int coordChunkZ = coord.chunkZ();
+                final int deltaX = Math.abs(originChunkX - coordChunkX);
+                final int deltaZ = Math.abs(originChunkZ - coordChunkZ);
+                return deltaX <= radius && deltaZ <= radius;
+            });
+        }
+
+        static <T extends Number> Condition<T> higherCondition(Property<T> property, T value) {
+            return __condition(property, (origin, t) -> t.doubleValue() > value.doubleValue());
+        }
+
+        static <T extends Number> Condition<T> higherEqualsCondition(Property<T> property, T value) {
+            return __condition(property, (origin, t) -> t.doubleValue() >= value.doubleValue());
+        }
+
+        static <T extends Number> Condition<T> lowerCondition(Property<T> property, T value) {
+            return __condition(property, (origin, t) -> t.doubleValue() < value.doubleValue());
+        }
+
+        static <T extends Number> Condition<T> lowerEqualsCondition(Property<T> property, T value) {
+            return __condition(property, (origin, t) -> t.doubleValue() <= value.doubleValue());
+        }
+
+        /**
+         * Allows any predicate which MUST be pure.
+         * No guarantee as to when called, and may not take advantage of future optimizations.
+         * <p>
+         * Use at your own risk.
+         */
+        @ApiStatus.Internal
+        static <T> Condition<T> __condition(Property<T> property, BiPredicate<Point, T> predicate) {
```

Comment:
should be named normally (`condition`, `userCondition`, `customCondition`?) + Experimental is probably better than Internal